### PR TITLE
add 2 examples : client_http_proxy and client_socks_proxy. fix #185 fix #436

### LINKED ISSUE #185: Document SOCKS and HTTP proxying mechanism
The current way to add proxying into node-mc-proto : 

``` js
var mc = require("minecraft-protocol");
var Http = require("http");
var proxy = { address: '192.168.1.1', port: 80 };
mc.Client.connect = function(port, host) {
  var self = this;
  var req = Http.request({
    host: proxy.address,
    port: proxy.port,
    method: 'CONNECT',
    path: host + ":" + port
  });
  req.on('connect', function(res, socket, head) {
    this.setSocket(socket);
  });
  req.end();
};
```

A better way to do this would be nice

### LINKED ISSUE #436: Socks proxy support error
So, i tried to make mc-proto work with socks proxy server. I started local socksv5 proxy, grabbed proxy.js from examples and changed properties on targetClient from this: 

``` js
var targetClient = mc.createClient({
    host: host,
    port: port,
    username: client.username,
    keepAlive:false,
    version:version
});
```

to this:

``` js
var targetClient = mc.createClient({
    stream: socks.createConnection({
      host: host,
      port: port
    }),
    username: client.username,
    keepAlive:false,
    version:version
  });
```

Mcproxy starts with no errors, but when im trying connect to it via vanilla client, i get error:

```
Incoming connection (127.0.0.1)
Connection error by server (127.0.0.1)  { [Error: Deserialization error for handshaking.toClient : Read error for name : 0 is not in the mappings value]
  field: 'handshaking.toClient',
  buffer: <Buffer 00 05 00 00 01> }
Error: Deserialization error for handshaking.toClient : Read error for name : 0 is not in the mappings value
    at ProtoDef.readMapper (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/datatypes/utils.js:43:34)
    at ProtoDef.read (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/protodef.js:102:31)
    at /home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:85:32
    at tryCatch (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/utils.js:83:12)
    at tryDoc (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/utils.js:90:10)
    at /home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:84:5
    at Array.forEach (native)
    at ProtoDef.readContainer (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:79:12)
    at ProtoDef.read (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/protodef.js:46:25)
    at ProtoDef.read (/home/user/proxy/node_modules/minecraft-protocol/node_modules/protodef/dist/protodef.js:102:31)
Connection closed by client (127.0.0.1)
```

I found similar error [here](http://www.gitterforum.com/discussion/prismarinejs-node-minecraft-protocol?page=3811)

Help?

### PR DESCRIPTION
**Pull request by rom1504**
Test it before merging

### CODE REVIEW

**Comment by rom1504** [comment]
might be good to fix https://github.com/PrismarineJS/node-minecraft-protocol/issues/481 too while we're at it

**Comment by NGTmeaty** [comment]
node client_http_proxy.js (censored)



events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:344:15)
    at Socket.socketOnEnd (_http_client.js:436:23)
    at emitNone (events.js:110:20)
    at Socket.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1047:12)
    at _combinedTickCallback (internal/process/next_tick.js:102:11)
    at process._tickCallback (internal/process/next_tick.js:161:9)
 



It seems to have a issue, if the connection is not ended before it is all sent, it will not work. The proxy was 100% working, as tested with my own home made script(in nodejs).

File: examples/client_http_proxy/client_http_proxy.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+const mc = require('minecraft-protocol');
+const Http = require("http");
+
+if(process.argv.length < 6 || process.argv.length > 8) {
+  console.log("Usage : node echo.js <host> <port> <proxyHost> <proxyPort> [<name>] [<password>]");
```
Comment (roblabla) [inline examples/client_http_proxy/client_http_proxy.js]: client_http_proxy.js, not echo.js

File: examples/client_socks_proxy/client_socks_proxy.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+const mc = require('minecraft-protocol');
+const Socks  = require("socks5-client");
+
+if(process.argv.length < 6 || process.argv.length > 8) {
+  console.log("Usage : node echo.js <host> <port> <proxyHost> <proxyPort> [<name>] [<password>]");
```
Comment (roblabla) [inline examples/client_socks_proxy/client_socks_proxy.js]: `client_socks_proxy.js`, not `echo.js`

**Comment by rom1504** [comment]
> It seems to have a issue, if the connection is not ended before it is all sent, it will not work

I don't understand that sentence.

When do you get that error ? At the connection ?

(I still didn't try these examples, planning to do it before merging)

**Comment by NGTmeaty** [comment]
@rom1504 The issue is right when it is connecting, and doesn't connect if the connection is not ended, and then if so, the proxy doesn't work.

**Comment by NGTmeaty** [comment]
Any ideas?

**Comment by rom1504** [comment]
ok trying to test this.

First step : what's a good simple to install http proxy

**Comment by rom1504** [comment]
trying https://github.com/TooTallNate/proxy

**Comment by NGTmeaty** [comment]
Squid proxy is easy.

**Comment by NGTmeaty** [comment]
Should also work.

**Comment by rom1504** [comment]
Ok I fixed a few things, http proxy tested to work.
To do : 
* [ ] add a test for the http proxy
* [ ] make sure socks5 also works
* [ ] add a test for the socks5 proxy

**Comment by NGTmeaty** [comment]
With the latest code, a error is spit out with a working proxy and account. 
```

examples/client_http_proxy/node_modules/minecraft-protocol/src/client/autoVersion.js:15
    if (err) throw err; // hmm
             ^

Error: connect ECONNREFUSED 127.0.0.1:25565
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:14)
```

**Comment by rom1504** [comment]
> Error: connect ECONNREFUSED 127.0.0.1:25565

localhost is relative to the proxy.

Are the proxy and the server running on the same machine ?

**Comment by NGTmeaty** [comment]
Nope, this is a external proxy.

**Comment by rom1504** [comment]
An external proxy cannot access /your/ localhost. You need to pass an external ip (or domain), not 127.0.0.1 as the server.

**Comment by NGTmeaty** [comment]
The server is also external.

**Comment by rom1504** [comment]
Ok then why are you giving `127.0.0.1:25565` as server configuration ?

**Comment by NGTmeaty** [comment]
I'm not.

**Comment by rom1504** [comment]
The command line is `node client_http_proxy.js <host> <port> <proxyHost> <proxyPort> [<name>] [<password>]`

What are you writing for this ?

**Comment by NGTmeaty** [comment]
[jake@JakeDesktop client_http_proxy]$ node client_http_proxy.js (minecraft server on the public internet) 25586 (proxy ip) 3128 (minecraft username) (minecraft password)

**Comment by rom1504** [comment]
@NGTmeaty are you sure your proxy supports CONNECT ?

**Comment by rom1504** [comment]
for example `149.56.97.249 80` works fine (see https://www.torvpn.com/en/proxy-list )

**Comment by NGTmeaty** [comment]
```
[jake@JakeDesktop client_http_proxy]$ node client_http_proxy.js mc.hypixel.net 25565 149.56.97.249 80  (censored)
/examples/client_http_proxy/node_modules/minecraft-protocol/src/client/autoVersion.js:15
    if (err) throw err; // hmm
             ^

Error: connect ECONNREFUSED 127.0.0.1:25565
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:14)

```

**Comment by rom1504** [comment]
Just tried that and it works fine here.
Are you using the last version of the code ? (did you `git pull`)
What version of node are you using ?

**Comment by NGTmeaty** [comment]
```
node --version
v8.1.3

```

Using that latest version from the pull request.

**Comment by rom1504** [comment]
`/examples/client_http_proxy/node_modules/minecraft-protocol/src/client/autoVersion.js`
That is weird
minecraft-protocol should be the parent dir, did you clone the repo ?

**Comment by NGTmeaty** [comment]
I did

**Comment by rom1504** [comment]
And why do you have a `client_http_proxy/node_modules/minecraft-protocol` then ?
do rm -rf client_http_proxy/node_modules

That example has no dependency.

**Comment by NGTmeaty** [comment]
Just redownloaded everything. [jake@JakeDesktop client_http_proxy]$ node client_http_proxy.js mc.hypixel.net 25565 149.56.97.249 80 (censored)
/home/jake/Desktop/node-minecraft-protocol/src/client/autoVersion.js:15
    if (err) throw err; // hmm
             ^

Error: connect ECONNREFUSED 127.0.0.1:25565
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:14)

**Comment by rom1504** [comment]
What operating system are you using ?

That's the only difference I can think of between what I'm doing and you.

I'm using linux

**Comment by NGTmeaty** [comment]
Linux here as well.

**Comment by rom1504** [comment]
can you add a console.log at https://github.com/PrismarineJS/node-minecraft-protocol/blob/6279ae9afe6679d23d02a9cf1160e89caac3e0a7/src/client/tcp_dns.js#L10 ?
it should go there.

Also it might be easier to talk at gitter.im/PrismarineJS/node-minecraft-protocol

**Comment by rom1504** [comment]
Ok the bug is ping isn't using the createClient options.

Fix is using them.

Problem is the connection get closed after ping, hence a new proxy connection needs to be opened.

So just passing a static socket stream to createClient won't work.

There are 2 ways to fix this : 
* override/hack the connect method to make it use the proxy thing
* add a customConnect options to createClient that will be called instead of the current connect whenever needed

second way is probably better.

**Comment by rom1504** [comment]
Did that. It works correctly now.

**Pull request merged**